### PR TITLE
Use CRC32 for file check

### PR DESCRIPTION
**Pull request by puzzledsab**
Related to #2396

### CODE REVIEW

File: sabnzbd/par2file.py [hunk 1]
```diff
@@ -106,40 +109,63 @@ def parse_par2_file(fname: str, md5of16k: Dict[bytes, str]) -> Tuple[str, Dict[s
     http://parchive.sourceforge.net/docs/specifications/parity-volume-spec/article-spec.html
     """
     total_size = os.path.getsize(fname)
+    metadata = {}
+    metadata["nr_files"] = -1
+    filedata = {}
```
Comment (Safihre) [inline sabnzbd/par2file.py:114]: Could you make `metadata` and `filedata` into `dataclass`? It's easier to read than dict's.

File: sabnzbd/utils/crc32calc.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/python3 -OO
+
+"""
+Functions for doing CRC32 calculations
+Written by animetoshi
```
Comment (Safihre) [inline sabnzbd/utils/crc32calc.py:5]: animetoshO 
;)

File: sabnzbd/assembler.py [hunk 3]
```diff
@@ -191,7 +188,10 @@ def assemble(nzo: NzbObject, nzf: NzbFile, file_done: bool):
                     # Could be empty in case nzo was deleted
                     if data:
                         fout.write(data)
-                        nzf.md5.update(data)
+                        if len(data) == nzo.article_size:
+                            nzf.crc32 = crc_multiply(nzf.crc32, nzo.crc32_coeff) ^ article.crc32
+                        else:
+                            nzf.crc32 = crc_concat(nzf.crc32, article.crc32, len(data))
```
Comment (Safihre) [inline sabnzbd/assembler.py:194]: Maybe we can have an `nzf.update_crc32(data, article.crc)` function instead of doing this here?

**Review by Safihre — CHANGES_REQUESTED** [review]
After these comments, do you prefer to merge as-is or first really get rid of the `md5` attributes etc?

**Comment by Safihre** [comment]
I update sabyenc3 to return the article-CRC instead of `crc_correct`.
It's part of this branch: https://github.com/sabnzbd/sabyenc/tree/openssl_link
I verified that it returns the same value as `zlib.crc32(raw_data)`.

Maybe it's easier to then base this PR against https://github.com/sabnzbd/sabnzbd/tree/feature/buffered_reader since it already uses this newer version of `sabyenc3`?

File: sabnzbd/par2file.py (hunk 1 above)
Comment (puzzledsab) [reply sabnzbd/par2file.py:113]: Sorry, I don't see how to get rid of the filedata dict? How to connect the CRC32 values to the right file?

**Comment by puzzledsab** [comment]
I'm not quite sure where to go with this patch. Have you considered keeping MD5 but making it optional? I think it should be good enough to be the only choice but I'm not 100% sure. The md5full value is even included in the database so I'm not sure about how much can break and if it needs some kind of special transition. I don't know enough git management stuff to know what kind of merging is preferable. I can submit a new PR to the other branch or I can wait until the ssl reader stuff is more settled. Whatever is easier.

**Comment by thezoggy** [comment]
Don't we used the md5 for dupe checking.. pretty sure we shouldn't use crc32 for that

**Comment by puzzledsab** [comment]
I looked at it a bit closer and the md5sum in the database seems to be the md5 of all the article IDs (nzo.md5sum), to identify duplicate nzbs. My patch only affects the file md5sum (nzf.md5sum), which is unrelated.

**Comment by Safihre** [comment]
> I looked at it a bit closer and the md5sum in the database seems to be the md5 of all the article IDs (nzo.md5sum), used to identify duplicate nzbs. My patch only affects the file md5sum (nzf.md5sum), which is unrelated.

Exactly!
So we can safely transition. If it's good enough for nzbget, it should be good enough for us, right?
Or you'd like to offer both? Speed vs reliability?
We could have a testing period with both methods and let users report the cases where CRC failed.
@animetosho can we really expect CRC to give false-positive results, so that it approves a download even though it's broken? I don't care if it has some false-negatives, if a few times in 1000 downloads it incorrectly triggers a full par2 scan.

**Comment by mnightingale** [comment]
> I update sabyenc3 to return the article-CRC instead of `crc_correct`.

The tests need updating :speak_no_evil:

Would there be any value in returning a custom type instead of a tuple?

**Comment by animetosho** [comment]
> can we really expect CRC to give false-positive results, so that it approves a download even though it's broken? I don't care if it has some false-negatives

All hash checks are subject to false-positives, and false-negatives should never occur.  
What generally matters is the *likelihood* of false-positives.

Assuming no 'extreme' scenarios (like someone deliberately trying to engineer a collision), it's practically impossible MD5 will false-positive, given a 128-bit hash size.  
CRC32 only has 32 bits, so the probability there is much higher.

I don't know how to calculate the probability, but if CRC32 is a 'perfect' 32-bit hash (I imagine it shouldn't be too far off), the probability of a false-positive should be somewhere around one in four billion - low, but not impossible.

**Comment by mnightingale** [comment]
I haven't looked at the implementation in detail but can or maybe it does already reduce the probability of collisions by checking multiple things?

Quick verify success = "CRC32 matches expected" and "All articles downloaded"
Otherwise do a full MD5 based check.

**Comment by puzzledsab** [comment]
I think it's supposed to keep a count of bad/missing articles, @Safihre ? I was thinking that we could also compare the file size with the expected value in the PAR2. The 16KB MD5 is already checked implicitly. With all those checks the chance of not discovering an error should be negligible.

**Comment by mnightingale** [comment]
> I was thinking that we could also compare the file size with the expected value in the PAR2.

I'd caution against using filesize as I think there is a good chance #2169 will land eventually :zipper_mouth_face:

**Comment by animetosho** [comment]
> Quick verify success = "CRC32 matches expected" and "All articles downloaded"  
> Otherwise do a full MD5 based check.

Unfortunately you won't know when an MD5 check is required, as the CRC32 check won't ever give you a false-negative.  
Of course, if you know that there's missing articles, hash checking isn't even needed.

File size is specified in a number of places:

* NZB by summing segment sizes (I think Sabnzbd primarily gets it from here?)
* yEnc headers
* PAR2

You could do a check between all of these, along with the total number of bytes decoded (file preallocation shouldn't hinder this).

**Comment by mnightingale** [comment]
> File size is specified in a number of places:
> 
> * NZB by summing segment sizes (I think Sabnzbd primarily gets it from here?)
> * yEnc headers
> * PAR2

You've  reminded me of a comment on #2396 about file offsets, where would be the best place to get the file offsets from if we wanted to do that approach?

**NZB**

```
<segment bytes="800000" number="1">...</segment>
<segment bytes="800000" number="2">...</segment>
```
Offsets derived from bytes/number

**yEnc headers**

```
=ypart begin=800001 end=1600000
```

---

I think SAB currently uses the NZB for the order to reconstruct the file so they are what should be used and perhaps it's possible that yEnc headers could lie or be constructed in a way to intentionally cause problems - such as tell the client this part is for position 1TB?

**Comment by Safihre** [comment]
In my previous attempt I just checked indeed the filesize on disk, seems plenty accurate for now:
https://github.com/sabnzbd/sabnzbd/commit/2ac08dd0e6659b24e283ca2d3f9f62963d82a565#diff-4d3d76bbdc64028f5e5c1442647035ee0b14a7baef4a1f48d7caffb9e9992a6fL2120

I also kept a check of bad-articles-per-NZF:
https://github.com/sabnzbd/sabnzbd/commit/2ac08dd0e6659b24e283ca2d3f9f62963d82a565#diff-4d3d76bbdc64028f5e5c1442647035ee0b14a7baef4a1f48d7caffb9e9992a6fL2115
The reason was that some indexers add fake files to their NZB's for tracking. These files are "missing", however, they are not part of any par2 set. So even if we let par2cmdline try to repair it, it can't anyway.
So we only care for missing articles if they are part of a par2 set.

If you don't mind I would suggest to close this PR in favor of #2407 ?

**Pull request closed without merging**